File: python/673. Number of Longest Increasing Subsequence.py
class Solution(object):
    def findNumberOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        dp = [1]*(n+1)
        cnt = [1]*(n+1)
        cnt[0] = dp[0] = 0
        for i in range(1, n+1):
            for j in range(1, i):
                if nums[j-1] < nums[i-1]:
                    if dp[i]<dp[j]+1:
                        dp[i] = dp[j]+1
                        cnt[i] = cnt[j]
                    elif dp[i]==dp[j]+1:
                        cnt[i] += cnt[j]

        res = 0
        max_val = max(dp)
        for i in range(n+1):
            if dp[i]==max_val:
                res += cnt[i]
        return res
        

File: go/src/leetcode/673. Number of Longest Increasing Subsequence/673. Number of Longest Increasing Subsequence.go
package main

import "fmt"

/**
思路：DP
f[i]表示以nums[i]结尾的最长递增子序列的长度
f[j] = max(f[j], f[i] + 1), nums[j] > nums[i], i < j
count[i]表示长度为f[i]的递增子序列的个数
 */
func findNumberOfLIS(nums []int) int {
	n := len(nums)
	f := make([]int, n)
	count := make([]int, n)
	for i := range nums {
		f[i] = 1
		count[i] = 1
	}

	var ans int
	maxVal := 1
	for i := 1; i < n; i++ {
		for j := 0; j < i; j++ {
			if nums[i] > nums[j] {
				if f[i] == f[j] + 1 {	//
					count[i] += count[j]
				} else if f[i] < f[j] + 1 {
					f[i] = f[j] + 1
					count[i] = count[j]
				}
			}
		}
		maxVal = max(maxVal, f[i])
	}

	fmt.Println(f, count, maxVal)
	for i := range f {
		if f[i] == maxVal {
			ans += count[i]
		}
	}

	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println(findNumberOfLIS([]int{1,3,5,4,7}))
	fmt.Println(findNumberOfLIS([]int{2,2,2,2,2}))
}
